Fixed issues around the data providers being called too many times.

Blackbaud-TrevorBurch opened the pull request.

Blackbaud-AlexKingman commented on src/app/public/modules/list-view-grid/list-view-grid.component.ts:
```diff
@@ -356,4 +375,10 @@ export class SkyListViewGridComponent
     return true;
   }
+
+  private arraysEqual(oldColumns: any[], newColumns: any[]) {
```
Nice private function here! Should these variables names be more generic since this could compare any array? Maybe just `arrayA` and `arrayB`?

Blackbaud-TrevorBurch replied:
Done

Blackbaud-AlexKingman commented on src/app/public/modules/list-view-grid/list-view-grid.component.ts:
```diff
@@ -157,7 +157,7 @@ export class SkyListViewGridComponent
 
     // Setup Observables for template
-    this.columns = this.gridState.map(s => s.columns.items).distinctUntilChanged();
+    this.columns = this.gridState.map(s => s.columns.items).distinctUntilChanged(this.arraysEqual);
```
Should this also have a `.takeUntil(this.ngUnsubscribe)`?

Blackbaud-TrevorBurch replied:
Done

Blackbaud-AlexKingman commented on src/app/public/modules/list-view-grid/list-view-grid.component.ts:
```diff
@@ -214,8 +214,8 @@ export class SkyListViewGridComponent
-      this.currentSearchText = this.state.map(s => s.search.searchText)
-        .distinctUntilChanged();
+    this.currentSearchText = this.state.map(s => s.search.searchText)
+      .distinctUntilChanged();
```
`.takeUntil(this.ngUnsubscribe)` ?

Blackbaud-TrevorBurch replied:
Done

Blackbaud-AlexKingman commented on src/app/public/modules/list-view-grid/list-view-grid.component.ts:
```diff
@@ -231,12 +242,18 @@ export class SkyListViewGridComponent
-        this.gridDispatcher.next(
-          new ListViewDisplayedGridColumnsLoadAction(displayedColumns, true)
-        );
+        this.selectedColumnIds
```
Just thinking out loud here... Is all this comparison logic too complicated for a `distinctUntilChanged()`? I see that you're fetching column Ids from an observable so it certainly is a bit more complex than a normal comparison. Just curious if you can think of a way to make this comparison check more consistent with the rest of our observables...

Blackbaud-TrevorBurch replied:
The key here is that we aren't comparing the old version of the observable to the new version we are comparing the new version of the observable to a local variable. 

However, looking at it again I do think that I can switch the order of the observables here so that if the if statement fails we don't even grab the value of the grid state.

Blackbaud-TrevorBurch replied:
Updated.

Blackbaud-SteveBrush commented:
@Blackbaud-TrevorBurch Is there an issue or detailed description we can add to the pull request to better explain the need for these changes?

Blackbaud-TrevorBurch commented:
@Blackbaud-SteveBrush https://github.com/blackbaud/skyux-list-builder/issues/12 I had linked it through Zenhub but didn't put it in the description.

There are two issues being fixed here in reality.

1. The grid state observable was firing out of order causing extra actions to be sent. I added a watch for the `lastUpdate` time to ensure this doesn't happen.
2. The grid was firing the column change event too many times and thus we were firing actions to often. I've fixed this in grid but also added a check here to ensure that we only fire an action to change the columns if it is actual necessary (i.e. I check to make sure the new id array is different from the previous one)

Blackbaud-TrevorBurch commented:
@Blackbaud-SteveBrush I added some unit tests for these changes
